File: advent-of-code/2023/cpp/3.cpp
#include <cstddef>
#include <utility>
#include <compare>
#include <concepts>

#include <span>
#include <vector>
#include <optional>
#include <string>
#include <string_view>
#include <unordered_map>

#include <iostream>
#include <fstream>

using Grid_view = std::span<std::string const>;

struct Coords {
    std::size_t x {};
    std::size_t y {};

    auto operator==(Coords const&) const -> bool = default;
};

template <>
struct std::hash<Coords> {
    auto operator()(Coords const coords) const -> std::size_t
    {
        using Hash = std::hash<std::size_t>;
        return Hash()(coords.x) + Hash()(coords.y);
    }
};

auto is_digit(char const c) -> bool
{
    return ('0' <= c) && (c <= '9');
}

auto adjacent_symbol_if(
    Grid_view const grid, Coords const coords, std::predicate<char> auto const& is_symbol)
    -> std::optional<Coords>
{
    for (int dy = -1; dy != 2; ++dy) {
        for (int dx = -1; dx != 2; ++dx) {
            std::size_t const x = coords.x + dx;
            std::size_t const y = coords.y + dy;
            if ((dx == 0 && dy == 0) || y >= grid.size() || x >= grid[y].size()) {
                continue;
            }
            if (is_symbol(grid[y][x])) {
                return Coords { x, y };
            }
        }
    }
    return std::nullopt;
}

auto parse_grid(
    Grid_view const                                 grid,
    std::predicate<char> auto const&                is_symbol,
    std::invocable<Coords, std::size_t> auto const& callback) -> void
{
    for (std::size_t y = 0; y != grid.size(); ++y) {
        std::string           digits;
        std::optional<Coords> adjacent_symbol;

        for (std::size_t x = 0; x != grid[y].size(); ++x) {
            char const c = grid[y][x];
            if (is_digit(c)) {
                digits.push_back(c);
                if (!adjacent_symbol) {
                    adjacent_symbol = adjacent_symbol_if(grid, { x, y }, is_symbol);
                }
            }
            else if (!digits.empty()) {
                if (adjacent_symbol) {
                    callback(adjacent_symbol.value(), std::stoull(digits));
                    adjacent_symbol.reset();
                }
                digits.clear();
            }
        }
        if (adjacent_symbol) {
            callback(adjacent_symbol.value(), std::stoull(digits));
        }
    }
}

auto part_one(Grid_view const grid) -> std::size_t
{
    std::size_t sum {};
    parse_grid(
        grid,
        [](char const c) { return !is_digit(c) && c != '.'; },
        [&](Coords, std::size_t const number) { sum += number; });
    return sum;
}

auto part_two(Grid_view const grid) -> std::size_t
{
    std::unordered_map<Coords, std::vector<std::size_t>> gear_map;
    parse_grid(
        grid,
        [](char const c) { return c == '*'; },
        [&](Coords const coords, std::size_t const number) { gear_map[coords].push_back(number); });

    std::size_t sum {};
    for (auto const& [_, numbers] : gear_map) {
        if (numbers.size() == 2) {
            sum += numbers.front() * numbers.back();
        }
    }
    return sum;
}

auto main() -> int
{
    std::vector<std::string> grid;
    {
        std::ifstream file { "inputs/3" };
        for (std::string line; std::getline(file, line);) {
            grid.push_back(std::move(line));
        }
    }
    std::cout << part_one(grid) << '\n' << part_two(grid) << '\n';
}

File: advent-of-code/2023/rust/3.rs
use std::collections::HashMap;

#[derive(PartialEq, Eq, Hash)]
struct Coords {
    x: usize,
    y: usize,
}

fn adjacent_symbol_if(
    grid: &[String],
    Coords { x, y }: Coords,
    is_symbol: impl Fn(u8) -> bool,
) -> Option<Coords> {
    for dy in -1..=1 {
        for dx in -1..=1 {
            if dy == 0 && dx == 0 {
                continue;
            }
            if let (Some(x), Some(y)) = (x.checked_add_signed(dx), y.checked_add_signed(dy)) {
                if let Some(row) = grid.get(y) {
                    if let Some(&char) = row.as_bytes().get(x) {
                        if is_symbol(char) {
                            return Some(Coords { x, y });
                        }
                    }
                }
            }
        }
    }
    None
}

fn parse_grid(
    grid: &[String],
    is_symbol: impl Fn(u8) -> bool,
    mut callback: impl FnMut(Coords, usize),
) {
    for y in 0..grid.len() {
        let mut digits = String::new();
        let mut adjacent_symbol: Option<Coords> = None;

        for x in 0..grid[y].len() {
            let char: u8 = grid[y].as_bytes()[x];
            if char.is_ascii_digit() {
                digits.push(char as char);
                if adjacent_symbol.is_none() {
                    adjacent_symbol = adjacent_symbol_if(grid, Coords { x, y }, &is_symbol);
                }
            }
            else if !digits.is_empty() {
                if let Some(coords) = adjacent_symbol {
                    callback(coords, digits.parse::<usize>().unwrap());
                    adjacent_symbol = None;
                }
                digits.clear();
            }
        }
        if let Some(coords) = adjacent_symbol {
            callback(coords, digits.parse::<usize>().unwrap());
        }
    }
}

fn part_one(grid: &[String]) -> usize {
    let mut sum = 0;
    parse_grid(grid, |c| (!c.is_ascii_digit() && c != b'.'), |_, number| sum += number);
    sum
}

fn part_two(grid: &[String]) -> usize {
    let mut gear_map = HashMap::<Coords, Vec<usize>>::new();
    parse_grid(
        grid,
        |c| c == b'*',
        |coords, number| gear_map.entry(coords).or_default().push(number),
    );
    gear_map
        .values()
        .filter_map(|vec| {
            if vec.len() == 2 {
                return Some(vec.first()? * vec.last()?);
            }
            None
        })
        .sum()
}

fn main() {
    let grid: Vec<String> = util::lines(3).collect();
    println!("1: {}", part_one(&grid));
    println!("2: {}", part_two(&grid));
}
